dask/dataframe/multi.py

```diff
@@ -419,6 +420,239 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
+    """
+    result = []
+
+    n, m = len(L) - 1, len(R) - 1
+    i, j = 0, -1
+    while R[j + 1] <= L[i]:
+        j += 1
+    J = []
+    while i < n:
+        partition = 0 if j < 0 else m - 1 if j >= m else j
+        lower = R[j] if j >= 0 and R[j] > L[i] else None
+        upper = R[j + 1] if j + 1 < m and (i == n - 1 or R[j + 1] < L[i + 1]) else None
+
+        J.append((partition, lower, upper))
+
+        i1 = i + 1 if j + 1 == m or (i + 1 < n and R[j + 1] >= L[i + 1]) else i
+        j1 = j + 1 if i + 1 == n or (j + 1 < m and L[i + 1] >= R[j + 1]) else j
+        if i1 > i:
+            result.append(J)
+            J = []
+        i, j = i1, j1
+
+    return result
+
+
+def merge_asof_padded(left, right, prev=None, next=None, **kwargs):
+    """ merge_asof but potentially adding rows to the beginning/end of right """
+    if left is None or right is None:
+        raise ValueError('dafuq did u do')
+
+    frames = []
+    if prev is not None:
+        frames.append(prev)
+    frames.append(right)
+    if next is not None:
+        frames.append(next)
+
+    return pd.merge_asof(left, pd.concat(frames), **kwargs)
+
+
+def merge_asof_indexed(left, right, **kwargs):
+    left = fix_overlap(left)
+    right = fix_overlap(right)
+
+    dsk = dict()
+    name = 'asof-join-indexed-' + tokenize(left, right, **kwargs)
+    meta = pd.merge_asof(left._meta_nonempty, right._meta_nonempty, **kwargs)
+
+    dependencies = [left, right]
+    tails = heads = None
+    if kwargs['direction'] in ['backward', 'nearest']:
+        tails = compute_tails(right, by=kwargs['right_by'])
+        dependencies.append(tails)
+    if kwargs['direction'] in ['forward', 'nearest']:
+        heads = compute_heads(right, by=kwargs['right_by'])
+        dependencies.append(heads)
+
+    for i, J in enumerate(pair_partitions(left.divisions, right.divisions)):
+        frames = []
+        for j, lower, upper in J:
+            slice = (methods.boundary_slice, (left._name, i),
+                     lower, upper, False)
+            tail = (tails._name, j) if tails is not None else None
+            head = (heads._name, j) if heads is not None else None
+            frames.append((apply, merge_asof_padded, [slice, (right._name, j),
+                           tail, head], kwargs))
+        dsk[(name, i)] = (pd.concat, frames)
+
+    graph = HighLevelGraph.from_collections(name, dsk,
+                                            dependencies=dependencies)
+    return new_dd_object(graph, name, meta, left.divisions)
+
+
+@wraps(pd.merge_asof)
+def merge_asof(left, right, on=None, left_on=None, right_on=None,
+               left_index=False, right_index=False, by=None, left_by=None,
+               right_by=None, suffixes=('_x', '_y'), tolerance=None,
+               allow_exact_matches=True, direction='backward'):
+    if direction not in ['backward', 'forward', 'nearest']:
+        raise ValueError("Invalid merge_asof direction. Choose from 'backward',\
```

Comment:
This will have weird spacing in the message. Something like

```
        raise ValueError("Invalid merge_asof direction. Choose from 'backward'"
                         " 'forward', or 'nearest'")


```

instead.